File: archive/future/python/providers/cloud/sandbox.py
"""Scalix Sandbox provider — wraps the existing scalix-sandbox SDK.

Provides Firecracker microVM isolation and GPU compute.
This is a thin wrapper that delegates to the existing scalix-sandbox package.

The scalix-sandbox SDK is already async, so we delegate directly.
"""

from __future__ import annotations

import logging
from typing import Any

from scalix.config import ScalixConfig, get_config
from scalix.exceptions import AuthenticationError, ProviderError, SandboxError
from scalix.providers.base import SandboxProvider
from scalix.types import SandboxResult

logger = logging.getLogger("scalix.providers.cloud.sandbox")

# Map SDK runtime names to scalix-sandbox runtime identifiers
RUNTIME_MAP = {
    "python": "python3.12",
    "python3.12": "python3.12",
    "python3.13": "python3.13",
    "node": "node22",
    "node22": "node22",
    "node24": "node24",
    "go": "go1.22",
    "go1.22": "go1.22",
    "rust": "rust-latest",
    "rust-latest": "rust-latest",
}

# Map SDK runtime names to the command used to execute code
RUNTIME_EXEC_CMD = {
    "python": ("python3", "-c"),
    "python3.12": ("python3", "-c"),
    "python3.13": ("python3", "-c"),
    "node": ("node", "-e"),
    "node22": ("node", "-e"),
    "node24": ("node", "-e"),
    "go": None,  # File-based
    "go1.22": None,
    "rust": None,  # File-based
    "rust-latest": None,
}


class ScalixSandboxProvider(SandboxProvider):
    """Execute code in Scalix Sandbox — Firecracker microVMs with GPU support.

    Wraps the existing `scalix-sandbox` Python SDK.
    Requires a Scalix API key (cloud mode).

    Advantages over DockerSandbox:
    - Firecracker microVM isolation (production-grade security)
    - GPU access (T4, A100, H100)
    - Persistent snapshots
    - Resource monitoring and metrics
    - Auto-scaling
    """

    def __init__(self, config: ScalixConfig | None = None) -> None:
        self._config = config or get_config()
        if not self._config.is_cloud_mode:
            raise AuthenticationError(
                "Scalix Sandbox requires an API key. "
                "Call scalix.configure(api_key='...') first."
            )
        self._sandbox_configured = False

    def _ensure_configured(self) -> None:
        """Ensure the scalix-sandbox SDK is configured with our API key."""
        if self._sandbox_configured:
            return

        try:
            from scalix_sandbox import configure
        except ImportError:
            raise ProviderError(
                "scalix-sandbox package not installed. "
                "Run: pip install scalix[cloud]"
            )

        configure(
            api_key=self._config.api_key,
            base_url=self._config.base_url,
        )
        self._sandbox_configured = True

    async def execute(
        self,
        code: str,
        runtime: str = "python",
        timeout: int = 30,
        gpu: str | None = None,
    ) -> SandboxResult:
        """Execute code in a Scalix Sandbox Firecracker microVM.

        Creates an ephemeral sandbox, runs the code, captures output,
        and cleans up. For long-running workloads, use the scalix-sandbox
        SDK directly.

        Args:
            code: Source code to execute.
            runtime: Language runtime ("python", "node", "go", "rust").
            timeout: Max execution time in seconds.
            gpu: GPU type ("t4", "a100", "h100") or None for CPU-only.
        """
        self._ensure_configured()

        from scalix_sandbox import Sandbox

        # Map runtime to scalix-sandbox format
        sandbox_runtime = RUNTIME_MAP.get(runtime)
        if not sandbox_runtime:
            raise SandboxError(
                f"Unsupported runtime '{runtime}'. "
                f"Supported: {', '.join(RUNTIME_MAP.keys())}"
            )

        # Build GPU config if requested
        gpu_config = None
        if gpu:
            from scalix_sandbox.types import GpuConfig
            gpu_config = GpuConfig(type=gpu, count=1)

        # Create sandbox
        logger.debug(
            "Creating Scalix Sandbox: runtime=%s, gpu=%s, timeout=%ds",
            sandbox_runtime, gpu, timeout,
        )

        sandbox = await Sandbox.create(
            runtime=sandbox_runtime,
            timeout_ms=timeout * 1000,
            gpu=gpu_config,
        )

        try:
            # Determine how to execute the code
            exec_cmd = RUNTIME_EXEC_CMD.get(runtime)

            if exec_cmd:
                # Inline execution (python -c, node -e)
                cmd, flag = exec_cmd
                result = await sandbox.run_command(cmd, [flag, code])
            else:
                # File-based execution (go, rust)
                ext_map = {"go": ".go", "go1.22": ".go", "rust": ".rs", "rust-latest": ".rs"}
                ext = ext_map.get(runtime, ".txt")
                file_path = f"/tmp/code{ext}"

                await sandbox.write_file(file_path, code)

                if runtime in ("go", "go1.22"):
                    result = await sandbox.run_command("go", ["run", file_path])
                elif runtime in ("rust", "rust-latest"):
                    result = await sandbox.run_command(
                        "sh", ["-c", f"rustc {file_path} -o /tmp/out && /tmp/out"]
                    )
                else:
                    result = await sandbox.run_command("cat", [file_path])

            # Convert to our SandboxResult format
            stdout = result.stdout() if callable(getattr(result, "stdout", None)) else str(getattr(result, "stdout", ""))
            stderr = result.stderr() if callable(getattr(result, "stderr", None)) else str(getattr(result, "stderr", ""))
            exit_code = getattr(result, "exit_code", 0)
            duration_ms = getattr(result, "duration_ms", None)

            return SandboxResult(
                stdout=stdout,
                stderr=stderr,
                exit_code=exit_code,
                duration_ms=duration_ms,
            )

        finally:
            # Always clean up the sandbox
            try:
                await sandbox.stop()
            except Exception as e:
                logger.warning("Failed to stop sandbox: %s", e)

File: archive/future/typescript/providers/cloud/sandbox.ts
/**
 * Scalix Sandbox provider — wraps the existing @scalix-world/sandbox SDK.
 *
 * Provides Firecracker microVM isolation and GPU compute.
 * This is a thin wrapper that delegates to the existing @scalix-world/sandbox package.
 */

import { getConfig, isCloudMode, dynamicImport, type ScalixConfig } from '../../config.js';
import { AuthenticationError, ProviderError, SandboxError } from '../../errors.js';
import type { SandboxProvider } from '../base.js';
import type { SandboxResult } from '../../types.js';

/** Map SDK runtime names to @scalix-world/sandbox runtime identifiers. */
const RUNTIME_MAP: Record<string, string> = {
  python: 'python3.12',
  'python3.12': 'python3.12',
  'python3.13': 'python3.13',
  node: 'node22',
  node22: 'node22',
  node24: 'node24',
  go: 'go1.22',
  'go1.22': 'go1.22',
  rust: 'rust-latest',
  'rust-latest': 'rust-latest',
};

/** Map runtime names to inline execution commands. null = file-based execution. */
const RUNTIME_EXEC_CMD: Record<string, [string, string] | null> = {
  python: ['python3', '-c'],
  'python3.12': ['python3', '-c'],
  'python3.13': ['python3', '-c'],
  node: ['node', '-e'],
  node22: ['node', '-e'],
  node24: ['node', '-e'],
  go: null,
  'go1.22': null,
  rust: null,
  'rust-latest': null,
};

/** File extensions for file-based runtimes. */
const EXT_MAP: Record<string, string> = {
  go: '.go',
  'go1.22': '.go',
  rust: '.rs',
  'rust-latest': '.rs',
};

/**
 * Execute code in Scalix Sandbox — Firecracker microVMs with GPU support.
 *
 * Wraps the existing `@scalix-world/sandbox` TypeScript SDK.
 * Requires a Scalix API key (cloud mode).
 *
 * Advantages over Docker sandbox:
 * - Firecracker microVM isolation (production-grade security)
 * - GPU access (T4, A100, H100)
 * - Persistent snapshots
 * - Resource monitoring and metrics
 * - Auto-scaling
 */
export class ScalixSandboxProvider implements SandboxProvider {
  private config: ScalixConfig;
  private configured = false;

  constructor(config?: ScalixConfig) {
    this.config = config ?? getConfig();
    if (!isCloudMode()) {
      throw new AuthenticationError(
        "Scalix Sandbox requires an API key. Call configure({ apiKey: 'sk_scalix_...' }) first.",
      );
    }
  }

  private async ensureConfigured(): Promise<void> {
    if (this.configured) return;

    try {
      const { configure } = await dynamicImport('@scalix-world/sandbox') as unknown as {
        configure: (opts: { apiKey: string; baseUrl?: string }) => void;
      };
      configure({
        apiKey: this.config.apiKey!,
        baseUrl: this.config.baseUrl,
      });
      this.configured = true;
    } catch {
      throw new ProviderError(
        '@scalix-world/sandbox package not installed. Run: npm install @scalix-world/sandbox',
      );
    }
  }

  async execute(
    code: string,
    options?: {
      runtime?: string;
      timeout?: number;
      gpu?: string;
    },
  ): Promise<SandboxResult> {
    await this.ensureConfigured();

    const runtime = options?.runtime ?? 'python';
    const timeout = options?.timeout ?? 30;
    const gpu = options?.gpu;

    // Validate runtime
    const sandboxRuntime = RUNTIME_MAP[runtime];
    if (!sandboxRuntime) {
      throw new SandboxError(
        `Unsupported runtime '${runtime}'. Supported: ${Object.keys(RUNTIME_MAP).join(', ')}`,
      );
    }

    const { Sandbox } = await dynamicImport('@scalix-world/sandbox') as unknown as {
      Sandbox: {
        create: (config: Record<string, unknown>) => Promise<SandboxInstance>;
      };
    };

    // Build creation config
    const createConfig: Record<string, unknown> = {
      runtime: sandboxRuntime,
      timeout: timeout * 1000,
    };

    if (gpu) {
      createConfig.gpu = { type: gpu, count: 1 };
    }

    const sandbox = await Sandbox.create(createConfig);

    try {
      const execCmd = RUNTIME_EXEC_CMD[runtime];
      let result: CommandResult;

      if (execCmd) {
        // Inline execution (python -c, node -e)
        const [cmd, flag] = execCmd;
        result = await sandbox.runCommand(cmd, [flag, code]);
      } else {
        // File-based execution (go, rust)
        const ext = EXT_MAP[runtime] ?? '.txt';
        const filePath = `/tmp/code${ext}`;

        await sandbox.writeFile(filePath, code);

        if (runtime === 'go' || runtime === 'go1.22') {
          result = await sandbox.runCommand('go', ['run', filePath]);
        } else if (runtime === 'rust' || runtime === 'rust-latest') {
          result = await sandbox.runCommand('sh', [
            '-c',
            `rustc ${filePath} -o /tmp/out && /tmp/out`,
          ]);
        } else {
          result = await sandbox.runCommand('cat', [filePath]);
        }
      }

      // Normalize output — the SDK may expose stdout/stderr as methods or properties
      const stdout =
        typeof result.stdout === 'function' ? result.stdout() : String(result.stdout ?? '');
      const stderr =
        typeof result.stderr === 'function' ? result.stderr() : String(result.stderr ?? '');
      const exitCode = result.exitCode ?? result.exit_code ?? 0;
      const durationMs = result.durationMs ?? result.duration_ms;

      return {
        stdout,
        stderr,
        exitCode,
        files: [],
        durationMs,
      };
    } finally {
      try {
        await sandbox.stop();
      } catch {
        // Best-effort cleanup
      }
    }
  }
}

/** Internal type for the @scalix-world/sandbox Sandbox instance. */
interface SandboxInstance {
  runCommand(
    cmd: string,
    args: string[],
    options?: Record<string, unknown>,
  ): Promise<CommandResult>;
  writeFile(path: string, content: string): Promise<void>;
  stop(): Promise<void>;
}

/** Internal type for command execution results. */
interface CommandResult {
  stdout: string | (() => string);
  stderr: string | (() => string);
  exitCode?: number;
  exit_code?: number;
  durationMs?: number;
  duration_ms?: number;
}
